src/se/vidstige/jadb/JadbDevice.java

```diff
@@ -181,16 +181,25 @@ public void push(File local, RemoteFile remote) throws IOException, JadbExceptio
 
     public void pull(RemoteFile remote, OutputStream destination) throws IOException, JadbException {
         Transport transport = getTransport();
-        SyncTransport sync = transport.startSync();
-        sync.send("RECV", remote.getPath());
-
-        sync.readChunksTo(destination);
+        try {
+        	SyncTransport sync = transport.startSync();
+        	sync.send("RECV", remote.getPath());
+        	sync.readChunksTo(destination);
+        }catch (Exception e) {
+        	if(transport != null) transport.close();
+        	throw e;
+		} 
```

Comment:
does this really work? Shouldn't we close both `sync` and `transport`.. 🤔

Reply:
Then tell why you did not close it bofere? I did not look into the implementation of the SyncTransport but yes it may be neccessary. I may have a look at it this evenening. You take more care about stream handling!

Reply:
SyncTransport does not have a close method -- that is a problem as well.
It seems it is used in two ways:
* creates it own Data{Input|Output}Streams (when called from Transport.startSync()) - in this case the streams should be closed (or Transport.close() should be called (the transport is opened for each command again and again)
* gets the DataStreams from AdbProtocolHandler. In this case it should not close them (they are alive for the whole server lifetime).

Solution is to create RawSyncTransport that gets datastreams from outside and does not close them (used directly by AdbProtocolHandler) and new SyncTransport that creates and closes them (used from Transport). Perhaps the new SyncTransport could keep reference to Transport and close it as well.

Reply:
I've implemented Closable for Transport and SyncTransport in #88 + #89. I suggest we resolve them first, and then this one will be easier.